src/Microsoft.ApplicationInsights/Extensibility/Implementation/ApplicationId/ProfileServiceWrapper.cs

```diff
@@ -0,0 +1,105 @@
+﻿namespace Microsoft.ApplicationInsights.Extensibility.Implementation.ApplicationId
+{
+    using System;
+    using System.Globalization;
+    using System.Net.Http;
+    using System.Threading.Tasks;
+    using Extensibility;
+
+    internal class ProfileServiceWrapper : IDisposable
+    {
+        internal readonly FailedRequestsManager FailedRequestsManager;
+
+        private HttpClient httpClient = new HttpClient();
+
+        internal ProfileServiceWrapper()
+        {
+            this.FailedRequestsManager = new FailedRequestsManager();
+        }
+
+        internal ProfileServiceWrapper(TimeSpan failedRequestRetryWaitTime)
+        {
+            this.FailedRequestsManager = new FailedRequestsManager(failedRequestRetryWaitTime);
+        }
+
+        public string ProfileQueryEndpoint { get; set; }
+
+        public async Task<string> FetchApplicationIdAsync(string instrumentationKey)
+        {
+            if (this.FailedRequestsManager.CanRetry(instrumentationKey))
+            {
+                try
+                {
+                    return await this.SendRequestAsync(instrumentationKey).ConfigureAwait(false);
+                }
+                catch (Exception ex)
+                {
+                    this.FailedRequestsManager.RegisterFetchFailure(instrumentationKey, ex);
+                    return null;
+                }
+            }
+            else
+            {
+                return null;
+            }
+        }
+
+        public void Dispose()
+        {
+            this.httpClient.Dispose();
+        }
+
+        /// <summary>Send HttpRequest to get config id.</summary>
+        /// <remarks>This method is internal so it can be moq-ed in a unit test.</remarks>
+        internal virtual async Task<HttpResponseMessage> GetAsync(string instrumentationKey)
```

Comment:
If you have an async method that has a simple "return await", then it should be transformed to a non-async, non-awaiting method. This avoids the async compiler-generated machinery (and the compiler doesn't do this for you).
i.e.
```C#
internal virtual Task<HttpResponseMessage> GetAsync(string instrumentationKey)
{
     Uri applicationIdEndpoint = this.GetApplicationIdEndPointUri(instrumentationKey.ToLowerInvariant());
     return this.httpClient.GetAsync(applicationIdEndpoint);
}
```

Reply:
Oh okay. Thanks for the tip!
Conventional wisdom was to sprinkle async/await everywhere and trust that the compiler would optimize. Can you recommend a source where I can learn more?